src/js/d3_svg_cell_renderer.js

```diff
@@ -0,0 +1,207 @@
+var utils = require('./utils');
+var $ = require('jquery');
+var _ = require('underscore');
+
+// TODO: handle accessing config properties cleaner
+
+function D3SVGRuleSet(renderer) {
+	var self = this;
+	self.rule_map = {};
+	self.renderer = renderer;
+
+	self.addRule = function(condition, d3_shape, attrs, z_index) {
+		var rule_id = Object.keys(self.rule_map).length;
+		if (z_index === undefined) {
+			z_index = rule_id;
+		}
+		self.rule_map[rule_id] = new D3SVGRule(rule_id, self.renderer, condition, d3_shape, attrs, z_index);
+		return rule_id;
+	};
+
+	self.removeRule = function(rule_id) {
+		delete self.rule_map[rule_id];
+	}
+
+	self.getRules = function() {
+		// returns a list of lists of rules to render in the given order
+		var z_map = {};
+		_.each(self.rule_map, function(rule, rule_id) {
+			z_map[rule.z_index] = z_map[rule.z_index] || [];
+			z_map[rule.z_index].push(rule);
+		});
+		return _.map(Object.keys(z_map).sort(), 
+			function(z) { return z_map[z];});
+	};
+
+	self.fromJSON = function(json_rules) {
+		self.rule_map = {};
+		_.each(json_rules, function(rule) {
+			self.addRule(rule.condition, rule.d3_shape, rule.attrs, rule.z_index);
+		});
+	};
+}
+
+function D3SVGRule(rule_id, renderer, condition, d3_shape, attrs, z_index) {
```

Comment:
1. Can you elaborate on what you mean by validation? And can you explain more how the code you linked to applies to rules?

2. My concern with organizing methods isn't for the compiler but for the human. 

I don't see managing state as unnecessary, I see it as natural - I think of oncoprints, tracks, and rules as stateful objects. They can be modified and manipulated.

3. Can you please clarify the structure you envision with regard to the various methods we might like to attach to rules, if we were to replace the rule class with a parameter map?